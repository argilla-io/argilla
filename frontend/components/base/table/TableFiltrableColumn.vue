<template>
  <div class="filter__container" v-if="column.filtrable">
    <button
      :data-title="column.tooltip"
      @click="openFilter(column)"
      :class="[visibleFilter || selectedOptions.length ? 'active' : '']"
    >
      <svgicon color="#4D4D4D" name="filter" width="16" />
      {{ column.name }}
    </button>
    <div class="table__filter" v-click-outside="close" v-if="visibleFilter">
      <select-options-search v-model="searchText" />
      <select-options
        ref="options"
        type="multiple"
        v-model="selectedOptions"
        :options="filterOptions(this.options, searchText)"
        :option-name="optionName"
        :option-counter="optionCounter"
      />
    </div>
  </div>
</template>

<script>
import "assets/icons/filter";
export default {
  props: {
    column: {
      type: Object,
      required: true,
    },
    filters: {
      type: Object,
      required: true,
    },
    data: {
      type: Array,
      required: true,
    },
  },
  data() {
    return {
      searchText: undefined,
      visibleFilter: false,
      selectedOptions: [],
    };
  },
  mounted() {
    this.$nextTick(() => {
      const filter = this.filters[this.column.field];
      if (filter) {
        if (filter.every(this.isObject)) {
          this.selectedOptions = this.options.filter((option) =>
            filter.some((f) => f.key === option.key && f.value === option.value)
          );
        } else {
          this.selectedOptions = filter;
        }
      }
    });
  },
  computed: {
    options() {
      const rawOptions = this.data.map((item) => item[this.column.field]);
      if (rawOptions.every((option) => this.isObject(option))) {
        const removedEmptyOptions = rawOptions
          .filter((opt) => Object.values(opt).length)
          .flatMap((opt) => opt);
        const optionsArray = Object.values(removedEmptyOptions).flatMap(
          (options) =>
            Object.keys(options).map((key) => {
              return { key: key, value: options[key] };
            })
        );
        const keys = ["key", "value"];
        return optionsArray.filter(
          (
            (s) => (o) =>
              ((k) => !s.has(k) && s.add(k))(keys.map((k) => o[k]).join("|"))
          )(new Set())
        );
      } else {
        return [...new Set(rawOptions)];
      }
    },
  },
  watch: {
    selectedOptions() {
      this.$emit("applyFilters", this.column, this.selectedOptions);
    },
    filters(val) {
      if (!Object.keys(val).length && this.selectedOptions.length) {
        this.selectedOptions = [];
      }
    },
  },
  methods: {
    isObject(obj) {
      return Object.prototype.toString.call(obj) === "[object Object]";
    },
    openFilter() {
      this.visibleFilter = true;
    },
    close() {
      this.visibleFilter = false;
    },
    filterOptions(options, text) {
      if (text === undefined) {
        return options;
      }
      const filtered = options.filter((id) =>
        JSON.stringify(id).toLowerCase().match(text.toLowerCase())
      );
      return filtered;
    },
    tableItemsCounter(option) {
      const keys = Object.keys(this.filters).filter(
        (k) => k !== this.column.field
      );
      const filteredData = this.data.filter((tableItem) => {
        return keys.every((key) => {
          if (this.filters[key].every((f) => this.isObject(f))) {
            return this.filters[key].find(
              (f) => f.value === tableItem[key][f.key]
            );
          } else {
            return this.filters[key].includes(tableItem[key]);
          }
        });
      });
      if (
        filteredData.every((data) => this.isObject(data[this.column.field]))
      ) {
        return filteredData.filter(
          (tableItem) =>
            tableItem[this.column.field][option.key] === option.value
        ).length;
      } else {
        return filteredData.filter(
          (tableItem) => tableItem[this.column.field] === option
        ).length;
      }
    },
    optionName(option) {
      return this.isObject(option) ? `${option.key}: ${option.value}` : option;
    },
    optionCounter(option) {
      return this.tableItemsCounter(option);
    },
  },
};
</script>
<style lang="scss" scoped>
.table__filter {
  background: $bg;
  position: absolute;
  top: 50px;
  left: -1em;
  margin-top: 0;
  padding: 10px 20px;
  z-index: 3;
  transform: translate(0);
  right: auto;
  min-width: 270px;
  border-radius: $border-radius;
  box-shadow: $shadow;
}

.filter {
  &__container {
    position: relative;
  }
  &__buttons {
    margin-top: 1em;
    text-align: right;
    display: flex;
    & > * {
      display: block !important;
      width: 100%;
      margin-right: $base-space;
      &:last-child {
        margin-right: 0;
      }
    }
  }
}
.filter-options {
  border: none;
  outline: none;
  height: 40px;
  background: transparent;
}
button {
  cursor: pointer;
  border: 0;
  outline: none;
  background: transparent;
  padding-left: 0;
  padding-right: 0;
  color: $black-87;
  display: flex;
  align-items: center;
  @include font-size(14px);
  @include media("<=desktop") {
    display: block;
    :deep(svg) {
      display: block;
      margin-right: 0 !important;
    }
  }
  &:hover,
  &.active {
    background: $black-4;
    min-height: 40px;
    padding: 0 1em;
    margin: 0 -1em;
    border-radius: $border-radius;
  }
  .svg-icon {
    margin-right: $base-space;
  }
}
</style>
