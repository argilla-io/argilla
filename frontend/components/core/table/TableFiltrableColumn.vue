<template>
  <div class="filter__container" v-if="column.filtrable">
    <button
      :data-title="column.tooltip"
      @click="openFilter(column)"
      :class="[visibleFilter || selectedOptions.length ? 'active' : '']"
    >
      {{ column.name }}
      <svgicon color="#4C4EA3" name="filtrable-column" />
    </button>
    <div class="table__filter" v-click-outside="close" v-if="visibleFilter">
      <input
        v-model="searchText"
        class="filter-options"
        type="text"
        autofocus
        placeholder="Search"
      />
      <ul>
        <li
          v-for="option in filterOptions(this.options, searchText)"
          :key="option"
        >
          <ReCheckbox
            :id="option"
            v-model="selectedOptions"
            class="re-checkbox--dark"
            :value="option"
          >
            {{ option }} ({{ datasetsCounter(option) | formatNumber }})
          </ReCheckbox>
        </li>
        <li
          v-if="!Object.entries(filterOptions(this.options, searchText)).length"
        >
          0 results
        </li>
      </ul>
    </div>
  </div>
</template>

<script>
import "assets/icons/filtrable-column";
export default {
  props: {
    column: {
      type: Object,
      required: true,
    },
    filters: {
      type: Object,
      required: true,
    },
    data: {
      type: Array,
      required: true,
    },
  },
  data() {
    return {
      searchText: undefined,
      visibleFilter: false,
      selectedOptions: [],
    };
  },
  mounted() {
    this.$nextTick(() => {
      if (this.filters[this.column.field]) {
        this.selectedOptions = this.filters[this.column.field];
      }
    });
  },
  computed: {
    options() {
      const rawOptions = this.data.map((item) => item[this.column.field]);
      return [...new Set(rawOptions)];
    },
  },
  watch: {
    selectedOptions() {
      this.$emit("applyFilters", this.column, this.selectedOptions);
    },
  },
  methods: {
    openFilter() {
      this.visibleFilter = true;
    },
    close() {
      this.visibleFilter = false;
    },
    filterOptions(options, text) {
      if (text === undefined) {
        return options;
      }
      let filtered = options.filter((id) =>
        id.toLowerCase().match(text.toLowerCase())
      );
      return filtered;
    },
    datasetsCounter(option) {
      const keys = Object.keys(this.filters).filter(
        (k) => k !== this.column.field
      );
      const filteredData = this.data.filter((dataset) => {
        return keys.every((key) => {
          return this.filters[key].includes(dataset[key]);
        });
      });
      return filteredData.filter(
        (dataset) => dataset[this.column.field] === option
      ).length;
    },
  },
};
</script>
<style lang="scss" scoped>
.table__filter {
  background: $bg;
  position: absolute;
  top: 50px;
  left: -1em;
  margin-top: 0;
  padding: 10px 20px;
  z-index: 3;
  transform: translate(0);
  right: auto;
  min-width: 270px;
  border-radius: $border-radius;
  box-shadow: $shadow;
  ul {
    list-style: none;
    max-height: 220px;
    overflow-y: auto;
    margin: 0 -1em;
    padding: 0 1em 1em;
  }
  li {
    padding: 0.4em 0;
  }
  .re-checkbox {
    margin: 0;
    width: 100% !important;
    cursor: default;
  }
}
.highlight-text {
  display: inline-block;
  // font-weight: 600;
  background: #ffbf00;
  line-height: 16px;
}

.filter {
  &__container {
    position: relative;
  }
  &__buttons {
    margin-top: 1em;
    text-align: right;
    display: flex;
    & > * {
      display: block !important;
      width: 100%;
      margin-right: 0.5em;
      min-height: 38px;
      &:last-child {
        margin-right: 0;
      }
    }
  }
}
.filter-options {
  border: none;
  outline: none;
  height: 40px;
  background: transparent;
}
button {
  cursor: pointer;
  border: 0;
  outline: none;
  background: transparent;
  padding-left: 0;
  color: $font-secondary;
  @include font-size(14px);
  font-family: $sff;
  &:hover,
  &.active {
    background: $bg;
    min-height: 40px;
    padding: 0 1em;
    margin: 0 -1em;
    border-radius: $border-radius;
    color: $primary-color;
    ::v-deep svg {
      margin-right: 0;
      & > * {
        fill: $primary-color !important;
      }
    }
  }
  .svg-icon {
    margin-left: 0.5em;
    margin-right: 1em;
  }
}
</style>
