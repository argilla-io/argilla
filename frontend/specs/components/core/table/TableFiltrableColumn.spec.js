import { mount } from "@vue/test-utils";
import TableFiltrableColumn from "@/components/core/table/TableFiltrableColumn";

function mountTableFiltrableColumn() {
  return mount(TableFiltrableColumn, {
    propsData: {
      filters: {
        owner: ["recognai"],
      },
      column: {
        class: "text",
        field: "owner",
        filtrable: "true",
        name: "Workspace",
        type: "text",
      },
      data: [
        {
          name: "dataset_1",
          owner: "recognai",
          task: "TokenClassification",
        },
      ],
    },
  });
}

describe("TableFiltrableColumn", () => {
  let spy = jest.spyOn(console, "error");
  afterEach(() => spy.mockReset());

  test("renders properly", () => {
    const wrapper = mountTableFiltrableColumn();
    expect(wrapper.html()).toMatchSnapshot();
  });
});
