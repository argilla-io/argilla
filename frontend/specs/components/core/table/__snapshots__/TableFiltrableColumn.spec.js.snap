// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableFiltrableColumn renders properly 1`] = `
<div class="filter__container"><button class="">
    Workspace
    <svg version="1.1" viewBox="0 0 15 11" class="svg-icon svg-fill">
      <path fill="#4C4EA3" stroke="none" pid="0" d="M.008.008v1.769h14.15V.008H.008zm2.527 6.064h9.096V4.304H2.535v1.768zm2.527 4.296h4.042V8.599H5.062v1.769z" _fill="#0508D9" fill-rule="nonzero"></path>
    </svg></button>
  <!---->
</div>
`;
